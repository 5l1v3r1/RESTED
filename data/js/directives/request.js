'use strict';

angular.module('RestedApp')
.directive('request', function($http) {
  return {
    restrict: 'E',
    templateUrl: 'views/directives/request.html',
    scope: {
      request: '='
    },
    link: function(scope, element, attrs, controllers) {
      scope.options = {
        methods: ['GET', 'POST', 'PUT', 'DELETE', 'PATCH', 'HEAD', 'JSONP']
      };

      // Set defaults
      scope.request = {
        method: scope.options.methods[0],
        url: 'http://www.vg.no',
        headers: {'Test-Header': 'hello'}
      };

      var processReturnData = function(response) {
        console.log(response);
        scope.response = response;
        scope.headers = response.headers();
      };

      scope.sendRequest = function() {
        $http(scope.request)
          .then(processReturnData, processReturnData);
      };

      scope.$watch('request', function(newVal, oldVal) {
        console.log('recieved Request:', newVal);
        if(newVal && newVal !== oldVal) {
          console.log('Request set to:', newVal);
          scope.request = newVal;
        }
      });
    }
  };
});
