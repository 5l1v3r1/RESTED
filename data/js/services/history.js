'use strict';

angular.module('RestedApp')
.factory('History', ['DEFAULT_HISTORY_SIZE', '$rootScope', '$filter', 'DB', 'Modal',
function(DEFAULT_HISTORY_SIZE, $rootScope, $filter, DB, Modal) {
  var errorHandler = function(event) {
    Modal.throwError('An error occured when reading/writing to indexedDB: ', event);
  };

  return {
    pushHistory: function(request) {
      try {
        if (!$rootScope.history) {
          $rootScope.history = [];
        }

        // Avoid spamming history: Only push new request
        if ($rootScope.history[0]
        && $rootScope.history[0].url === request.url
        && $rootScope.history[0].method === request.method) {
          return;
        }

        // "Push" the request (actually unshift to the beginning)
        $rootScope.history.unshift(request);

        // Truncate the list to historySize
        var historySize = $rootScope.options.historySize != null
          ? $rootScope.options.historySize
          : DEFAULT_HISTORY_SIZE;
        $rootScope.history = $rootScope.history.slice(0, historySize);

        DB.history.set({name: 'history', requests: $rootScope.history})
          .then(Modal.remove, errorHandler);

      } catch (e) {
        console.error(e);
        Modal.throwError('error in history.pushHistory: ', e.toString());
      }
    },
    // removeRequestFromHistory: function(collection, index, collectionIndex) {
    //   Modal.set({
    //     title: 'Confirm deletion',
    //     body: 'Please confirm you wish to remove this request from your saved collection',
    //     actions: [{
    //       text: 'Confirm',
    //       click: function() {
    //         collection.requests.splice(index, 1);
    //         DB.collections.set(collection).then(Modal.remove, errorHandler);
    //       }
    //     }]
    //   });
    // },
    // clearCollection: function(callback) {
    //   Modal.set({
    //     title: 'Confirm deletion',
    //     body: 'Please confirm that you wish to delete your entire collection. This is not reversible',
    //     actions: [{
    //       text: 'Confirm',
    //       click: function() {
    //         if (!$rootScope.collections || !$rootScope.collections[0]) {
    //           return callback();
    //         }
    //         $rootScope.collections[0].requests = [];
    //         DB.collections.set($rootScope.collections[0]).then(Modal.remove, errorHandler);
    //         callback();
    //       }
    //     }]
    //   });
    // }
  };
}]);

