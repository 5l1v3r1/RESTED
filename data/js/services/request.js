'use strict';

angular.module('RestedApp')
.factory('Request', function() {

  /**
  * Prepend http:// if missing from url
  * Make sure to allow https
  */
  var prependHttp = function(url) {
    if(!(/^https?:\/\//.test(url))) {
      url = 'http://' + url;
    }
    return url;
  };

  /**
  * Takes an url in the form of site.com/{{key}} and
  * maps the key up against a variable in the params
  * object. It then returns the complete URL.
  *
  * The function also collapses {{key}} to nothing
  * if key is missing, and removes any double slashes
  * left behind.
  */
  var mapParameters = function(url, params) {
    return url.replace(/\{\{(\w+)\}\}/g, function(match, capture) {
      var param = params ? params[capture] : null;
      return param ? param : '';
    }).replace(/([^:]\/)\/+/g, '$1');
  };

  var createXMLHttpRequest = function(req) {
    var request = new XMLHttpRequest(true, true);
    request.open(req.method, req.url);

    if(Array.isArray(req.headers)) {
      req.headers.forEach(function(header) {
        if(header.name) {
          request.setRequestHeader(header.name, header.value);
        }
      });
    }

    return request;
  };

  return {
    run: function(request, parameters, fn) {
      var requestCopy = angular.copy(request);
      var url = prependHttp(requestCopy.url);
      url = mapParameters(url, parameters);

      requestCopy.url = url;
      var req = createXMLHttpRequest(requestCopy);
      req.onloadend = fn.bind(req);

      req.send(request.data);
    },
    /* For unit tests only */
    _prependHttp: prependHttp,
    _mapParameters: mapParameters,
    _createXMLHttpRequest: createXMLHttpRequest
  };
});

