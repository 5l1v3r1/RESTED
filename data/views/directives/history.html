<div class="panel-heading">
  <div class="pull-right hoverhand">
    <slidey-button data-config="clearHistoryConfig"
                   data-click-event="clearHistory()">
    </slidey-button>
  </div>
</div>
<div class="panel-body">
  <ul class="panel-body collection-body"
      data-ng-hide="history.length === 0">
    <li data-ng-repeat="request in history">

      <div class="list-group">
        <a href="#"
           class="list-group-item"
           data-ng-class="{active: request == selectedRequest}"
           data-ng-click="selectRequest(request)">
          <h4 class="list-group-item-heading">
            {{request.method}}
            <div class="pull-right"
                 id="removeRequest">
              <slidey-button data-config="removeRequestConfig"
                             data-click-event="removeFromHistory($index)">
              </slidey-button>
            </div>
          </h4>

          <p class="list-group-item-text">
            {{request.url}}
          </p>
        </a>
      </div>
    </li>
  </ul>
  <div data-ng-show="!history || history.length === 0">
    <h5>
        You have no recorded history. Send some requests and start making your legacy!
    </h5>
  </div>
</div>
