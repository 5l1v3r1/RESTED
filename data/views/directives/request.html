<div>
  <section class="panel panel-default">
    <div class="panel-heading">
      Request

      <div class="pull-right hoverhand">
        <div data-ng-click="addRequest(request)"
             class="fa fa-plus"
             title="Add request to collection">
        </div>
      </div>
    </div>
    <div class="panel-body">
      <form data-ng-submit="sendRequest()"
            class="form-horizontal">

        <div class="form-group">
          <label for="url"
                 class="col-sm-2 control-label">
                 URL
          </label>
          <div class="col-sm-10">
            <input type="text"
                   class="form-control"
                   id="url"
                   placeholder="{{::getRandomURL()}}"
                   data-ng-model="request.url"
            />
          </div>
        </div>

        <div class="form-group">
          <label for="method"
                 class="col-sm-2 control-label">
                 Method
          </label>

          <div class="col-sm-7">
            <select class="form-control"
                    id="method"
                    data-ng-model="request.method"
                    data-ng-options="method for method in options.methods">
            </select>
          </div>
          <div class="col-sm-3">
            <input type="submit"
                   class="btn btn-primary"
                   value="Send request"
            />
          </div>
        </div>

        <div class="hoverhand"
             data-ng-click="slideToggle('headerSlider')"
             title="Show headers">
          <h4>
            Headers
            <i class="fa fa-angle-right animate"
                data-ng-class="{'fa-rotate-90': !!slidden.headerSlider}">
            </i>
          </h4>
        </div>

        <div id="headerSlider" class="row start-hidden">
          <div class="header-item"
               data-ng-repeat="header in headers">
            <div class="form-group col-xs-6">
              <input type="text"
                     class="form-control"
                     data-ng-model="header.name"
                     placeholder="Name"
              />
            </div>
            <div class="form-group col-xs-6">
              <input type="text"
                     class="form-control"
                     data-ng-model="header.value"
                     placeholder="Value"
              />
            </div>
          </div>
          <h5 class="col-xs-12 hoverhand"
              data-ng-click="addHeader()"
              title="Add header">
            <i class="fa fa-plus"></i>
            Add header
          </h5>
        </div>

        <!-- POST, PUT, PATCH get data -->
        <div class="hoverhand"
             data-ng-click="slideToggle('bodySlider')"
             title="Show request body">
          <h4>
            Request body
            <i class="fa fa-angle-right animate"
                data-ng-class="{'fa-rotate-90': !!slidden.bodySlider}">
            </i>
          </h4>
        </div>

        <div id="bodySlider" class="row start-hidden">
          <div class="col-xs-12">
            <textarea class="form-control"
                      rows="10"
                      data-ng-model="request.data">
            </textarea>
          </div>
        </div>
      </form>
    </div>
  </section>

  <section data-ng-if="response" class="panel panel-default">
    <div class="panel-heading one-line-text">
      Response - {{response.config.url}}
    </div>
    <div class="panel-body">
      <h2 class="col-xs-12" data-ng-style="{color: (response.status | status)}">
        {{response.status}} <small>{{response.statusText}}</small>
      </h2>

      <content-slider class="col-xs-12"
                      title="Headers"
                      content="response.headers"
                      start-hidden="true">
      </content-slider>

      <content-slider class="col-xs-12"
                      title="Response body"
                      content="response.data">
      </content-slider>
    </div>
  </section>
</div>
