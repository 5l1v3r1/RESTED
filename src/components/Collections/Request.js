import React from 'react';
import { connect } from 'react-redux';
import { DragSource, DropTarget } from 'react-dnd';
import { ListGroup } from 'react-bootstrap';
import flow from 'lodash.flow';

import IconButton from 'components/IconButton';
import * as Actions from 'store/collections/actions';

import { StyledRequest, AsideButtons, MainContent } from './StyledComponents';
import * as Type from './dropTypes';

/**
 * Specifies the drag source contract.
 * Only `beginDrag` function is required.
 */
const requestSource = {
  canDrag() {
    // TODO Disallow drag if editing name
    return true;
  },

  beginDrag({ id, index, collectionIndex }) {
    return { id, index, collectionIndex };
  },
};

const requestTarget = {
  hover(props, monitor) {
    const dragIndex = monitor.getItem().index;
    const dragCollectionIndex = monitor.getItem().collectionIndex;
    const dragUUID = monitor.getItem().id;
    const hoverIndex = props.index;
    const hoverCollectionIndex = props.collectionIndex;
    const hoverUUID = props.id;

    // Abort further processing if dragged item is hovering over itself
    if (dragUUID === hoverUUID) {
      return;
    }

    // Note: we're mutating the monitor item here.
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    /* eslint-disable no-param-reassign */
    monitor.getItem().index = hoverIndex;
    monitor.getItem().collectionIndex = hoverCollectionIndex;
    /* eslint-enable no-param-reassign */

    // Update redux orders
    props.reorderRequest({
      // Source
      collectionIndex: dragCollectionIndex,
      requestIndex: dragIndex,
    }, {
      // Target
      collectionIndex: hoverCollectionIndex,
      requestIndex: hoverIndex,
    });
  },
};

function Request(props) {
  const {
    connectDragSource,
    connectDropTarget,
    isDragging,
    id,
    collectionIndex,
    method,
    url,
    deleteRequest,
  } = props;

  return connectDragSource(connectDropTarget(
    <div> {/* Need a wrapper div for React DnD support */}
      <StyledRequest isDragging={isDragging}>
        <ListGroup componentClass="div">
          <div className="list-group-item">
            <AsideButtons>
              <IconButton
                tooltip="Edit"
                icon="cog"
              />
              <IconButton
                tooltip="Delete"
                icon="trash"
                onClick={() => deleteRequest(id, collectionIndex)}
              />
            </AsideButtons>

            <MainContent>
              <h4>{method}</h4>
              {url}
            </MainContent>
          </div>
        </ListGroup>
      </StyledRequest>
    </div>,
  ));
}

export default flow(
  DropTarget(Type.Request, requestTarget, connector => ({
    connectDropTarget: connector.dropTarget(),
  })),
  DragSource(Type.Request, requestSource, (connector, monitor) => ({
    connectDragSource: connector.dragSource(),
    isDragging: monitor.isDragging(),
  })),
  connect(null, Actions),
)(Request);

