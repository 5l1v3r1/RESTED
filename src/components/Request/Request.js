import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field, Fields, FieldArray, propTypes } from 'redux-form';
import { Panel } from 'react-bootstrap';

import URLField from './URLField';
import MethodField from './MethodField';
import HeadersField from './HeadersField';
import BasicAuthField from './BasicAuthField';
import BodyField from './BodyField';

import * as Actions from '../../store/request/actions';
import { DEFAULT_REQUEST } from '../../constants/constants';

function Titlebar() {
  return (
    <h2>
      Request
    </h2>
  );
}

function Request({ placeholderUrl, handleSubmit, sendRequest }) {
  return (
    <Panel header={<Titlebar />}>
      <form onSubmit={handleSubmit(sendRequest)}>
        <Field
          name="url"
          component={URLField}
          placeholderUrl={placeholderUrl}
        />
        <Field
          name="method"
          component={MethodField}
        />
        <FieldArray
          name="headers"
          component={HeadersField}
        />
        <Fields
          names={['basicAuth.username', 'basicAuth.password']}
          component={BasicAuthField}
        />
        <BodyField />
      </form>
    </Panel>
  );
}

Request.propTypes = {
  placeholderUrl: PropTypes.string,
  ...propTypes,
};

const formOptions = {
  form: 'request',
};

const mapStateToProps = ({ request: { useFormData, placeholderUrl } }) => ({
  useFormData,
  placeholderUrl,
  initialValues: DEFAULT_REQUEST,
});

export { Request };

// TODO this
/* eslint-disable no-func-assign */
Request = reduxForm(formOptions)(Request);
Request = connect(mapStateToProps, Actions)(Request);
/* eslint-enable no-func-assign */
export default Request;

