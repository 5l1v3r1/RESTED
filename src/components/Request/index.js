import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field, Fields, FieldArray, propTypes } from 'redux-form';
import { Panel, Form } from 'react-bootstrap';
import flow from 'lodash.flow';

import * as requestActions from 'store/request/actions';
import requestValidation from 'utils/requestValidation';
import { DEFAULT_REQUEST } from 'constants/constants';

import Titlebar from './Titlebar';
import URLField from './URLField';
import MethodField from './MethodField';
import HeadersField from './HeadersField';
import BasicAuthField from './BasicAuthField';
import BodyField from './BodyField';

export const requestForm = 'request';

function Request({ placeholderUrl, handleSubmit, sendRequest }) {
  return (
    <Panel header={<Titlebar />}>
      <Form horizontal onSubmit={handleSubmit(sendRequest)}>
        <Field
          name="url"
          component={URLField}
          placeholderUrl={placeholderUrl}
        />
        <Field
          name="method"
          component={MethodField}
        />
        <FieldArray
          name="headers"
          component={HeadersField}
        />
        <Fields
          names={['basicAuth.username', 'basicAuth.password']}
          component={BasicAuthField}
        />
        <BodyField />
      </Form>
    </Panel>
  );
}

Request.propTypes = {
  placeholderUrl: PropTypes.string,
  ...propTypes,
};

const formOptions = {
  form: requestForm,
  validate: requestValidation,
};

const mapStateToProps = ({ request: { useFormData, placeholderUrl } }) => ({
  useFormData,
  placeholderUrl,
  initialValues: DEFAULT_REQUEST,
});

export { Request };

export default flow(
  reduxForm(formOptions),
  connect(mapStateToProps, requestActions),
)(Request);

