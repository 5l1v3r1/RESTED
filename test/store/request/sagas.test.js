import Immutable from 'immutable';
import { put, call, apply, select } from 'redux-saga/effects';
import { change } from 'redux-form';
import 'whatwg-fetch';

import { requestForm } from 'components/Request';
import { fetchData, createResource, buildHeaders, getParameters, getUrl, getBeforeTime, getMillisPassed } from 'store/request/sagas';
import { getPlaceholderUrl } from 'store/request/selectors';
import { pushHistory } from 'store/history/actions';
import * as types from 'store/request/types';
import { prependHttp } from 'utils/request';

const mockRequest = {
  method: 'GET',
  url: 'http://visitnorway.com',
  headers: [{
    name: 'Foo',
    value: 'Bar',
  }],
  formData: [{
    name: 'Yes',
    value: 'Sir',
  }],
};

describe('fetchData saga', () => {
  const mockHeaders = new Headers({
    Foo: 'Bar',
  });
  const mockResponseData = {
    status: 200,
    statusText: 'SuperSmashingGreat!',
    headers: new Headers({
      Header: 'Value',
    }),
  };
  const mockResponse = new Response(new Blob(), mockResponseData);

  const iterator = fetchData({ request: mockRequest });

  it('should dispatch an executeRequest action', () => {
    expect(iterator.next().value).toEqual(put({
      type: types.EXECUTE_REQUEST,
    }));
  });

  it('should call createResource to build a URL', () => {
    expect(iterator.next().value).toEqual(
      call(createResource, mockRequest),
    );
  });

  it('should call buildHeaders to prepare the header set', () => {
    expect(iterator.next('foo').value).toEqual(
      call(buildHeaders, mockRequest),
    );
  });

  it('should push the history', () => {
    expect(iterator.next(mockHeaders).value).toEqual(
      put(pushHistory(Immutable.fromJS(mockRequest)
        .set('url', 'foo')
        .set('id', 'test-UUID'),
      )),
    );
  });

  let timeBefore;
  it('should calculate the time passed', () => {
    timeBefore = Date.now();
    expect(iterator.next().value).toEqual(
      call(getBeforeTime),
    );
  });

  it('should fetch the resource', () => {
    const body = new FormData();
    body.append('Yes', 'Sir');

    expect(iterator.next(timeBefore).value).toEqual(
      call(fetch, 'foo', {
        method: 'GET',
        body,
        redirect: 'follow',
        headers: new Headers({
          Foo: 'Bar',
        }),
        credentials: 'include',
      }),
    );
  });

  const timePassed = 1337;
  it('should calculate the time passed', () => {
    expect(iterator.next(mockResponse).value).toEqual(
      call(getMillisPassed, timeBefore),
    );
  });

  it('should get the response text', () => {
    expect(iterator.next(timePassed).value).toEqual(
      apply(mockResponse, mockResponse.text),
    );
  });

  it('should dispatch a receiveResponse action', () => {
    expect(iterator.next().value).toEqual(put({
      type: types.RECEIVE_RESPONSE,
      response: {
        method: 'GET',
        time: timePassed,
        url: mockResponse.url,
        body: undefined,
        status: mockResponse.status,
        statusText: mockResponse.statusText,
        headers: [{
          name: 'header',
          value: 'Value',
        }],
      },
    }));
  });

  it('should dispatch a requestFailed when something throws', () => {
    const iterator2 = fetchData({ request: mockRequest });
    const error = { foo: 'bar' };
    iterator2.next();

    expect(iterator2.throw(error).value).toEqual(
      put({ type: types.REQUEST_FAILED, error }),
    );
  });

  it('should be done', () => {
    expect(iterator.next().done).toEqual(true);
  });
});

describe('createResource saga', () => {
  const iterator = createResource(mockRequest);
  const mockUrl = 'foo.com/{{foo}}';
  const mockUrlVariables = {
    foo: 'bar',
  };

  it('should call getUrl with the request', () => {
    expect(iterator.next().value).toEqual(
      call(getUrl, mockRequest),
    );
  });

  it('should call getParameters to get the urlVariables', () => {
    expect(iterator.next(mockUrl).value).toEqual(
      call(getParameters),
    );
  });

  it('should call prependHttp on the url', () => {
    expect(iterator.next(mockUrlVariables).value).toEqual(
      call(prependHttp, 'foo.com/bar'),
    );
  });

  it('should return the resulting resource', () => {
    expect(iterator.next('http://foo.com/bar').value).toBe(
      'http://foo.com/bar',
    );
  });

  it('should be done', () => {
    expect(iterator.next().done).toEqual(true);
  });
});

describe('getUrl saga', () => {
  const iterator = getUrl({});
  const fallbackUrl = 'http://test.com';

  it('should return the URL on the request if present', () => {
    const iterator2 = getUrl(mockRequest);
    expect(iterator2.next().value).toBe(mockRequest.url);
  });

  it('should getch the placeholderUrl from the store', () => {
    expect(iterator.next().value).toEqual(
      select(getPlaceholderUrl),
    );
  });

  it('should dispatch a URL change event to redux form', () => {
    expect(iterator.next(fallbackUrl).value).toEqual(
      put(change(requestForm, 'url', fallbackUrl)),
    );
  });

  it('should return the fallbackUrl', () => {
    expect(iterator.next().value).toBe(fallbackUrl);
  });

  it('should be done', () => {
    expect(iterator.next().done).toEqual(true);
  });
});

