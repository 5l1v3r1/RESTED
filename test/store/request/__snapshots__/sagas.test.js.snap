// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changeBodyTypeSaga saga when bodyType is anything else should throw 1`] = `"Body type foo is not supported"`;
